src/connectors/controller/index.ts

```diff
@@ -0,0 +1,108 @@
+
+import Controller, { type ControllerOptions } from "@cartridge/controller";
+import { Connector, type ConnectArgs, type ConnectorData } from "../connector"
+import { ConnectorNotConnectedError, UserNotConnectedError, UserRejectedRequestError } from "../../errors"
+
+import { CONTROLLER_ICON, RPC_URLS } from "./constants";
+
+export class ControllerConnector extends Connector {
+  private controller: Controller | null;
+
+  constructor(options: Partial<ControllerOptions> = {}) {
+    super();
+
+    // Can pass in defaultChainId = SN_SEPOLIA to connect to Sepolia testnet
+    options.defaultChainId = options.defaultChainId || constants.StarknetChainId.SN_MAIN;
+    options.chains = options.chains || [{ rpcUrl: RPC_URLS[options.defaultChainId as keyof typeof RPC_URLS] }];
+
+    this.controller = this.available()
+      ? new Controller(options as ControllerOptions)
+      : null;
+  }
+
+  get id() { return "cartridgeController" }
+  get name() { return "Cartridge Controller" }
+  get icon() { return { light: CONTROLLER_ICON, dark: CONTROLLER_ICON } }
+
+  available() {
+    return typeof window !== "undefined";
+  }
+
+  async ready() {
+    if (!this.controller) {
+      return false;
+    }
+
+    const account = await this.controller.probe();
+    return account !== null;
+  }
+
+  async connect(_args?: ConnectArgs): Promise<ConnectorData> {
+    if (!this.controller) { throw new ConnectorNotConnectedError(); }
+
+    const account = await this.controller.connect();
+
+    if (!account) { throw new UserNotConnectedError(); }
+
+    const chainId = await this.chainId();
+
+    /**
+     * Taken from src/connectors/injected/index.ts
+     * @dev This emit ensures compatibility with starknet-react
+     */
+    this.emit("connect", { account: account.address, chainId })
+
+    return {
+      account: account.address,
+      chainId: chainId
+    };
+  }
+
+  async disconnect(): Promise<void> {
+    if (!this.controller) { throw new ConnectorNotConnectedError(); }
+
+    return this.controller.disconnect();
+  }
+
+  async account(_provider: ProviderOptions | ProviderInterface): Promise<AccountInterface> {
+    if (!this.controller) { throw new ConnectorNotConnectedError(); }
+
+    const account = await this.controller.probe();
+
+    if (!account) { throw new UserNotConnectedError(); }
+
+    return account;
+  }
+
+  async chainId(): Promise<bigint> {
+    if (!this.controller) { throw new ConnectorNotConnectedError(); }
+
+    const account = await this.controller.probe();
+
+    if (!account) { throw new UserNotConnectedError(); }
+
+    return BigInt(await account.getChainId());
+  }
+
+  async request<T extends RpcMessage["type"]>(
+    call: RequestFnCall<T>
+  ): Promise<RpcTypeToMessageMap[T]["result"]> {
+    if (!this.controller) { throw new ConnectorNotConnectedError(); }
+
+    try {
+      return await this.controller.request(call);
```

Comment:
I'm having some issues testing this locally at this line -- signing message or opening sessions all fail here.

When attempting to sign a `Hello World` message, I get:

```
Error: Invalid message provided: Type StarknetDomain not found
```

When attempting to open a session, I get:

```
Objects are not valid as a React child (found: object with keys {Contract Address, selector}). If you meant to render a collection of children, use an array instead.
```

This implementation of `request` follows closely the implementation of the other connectors. Does anyone with more experience with Starknet wallets have insight? I'm guessing it's something simple.

Reply:
The situation has been resolved. The first error was due to SNIP-12 compatibility issues, which have been documented. The second error is a UI issue with the upstream Controller and is the result of using unsupported wallet functionality.